Propagate Connection Closed Information up to top-level (fix #465)

sebsto opened the pull request:
This PR implements a mechanism to propagate connection loss information from the Lambda runtime client to the runtime loop, enabling termination without backtrace when the connection to the Lambda control plane (or a Mock Server) is lost.

The changes are:

- When the connection is lost, `ChannelHandlerDelegate.channelInnactive()` now correctly calls `resume(throwing:)` on the ending continuation, for all states (`.waitingForNextInvocation ` and `.sentResponse`). This eliminates the hangs on connection lost..

- I added top-level error handling on `LambdaRuntime._run()` 

- Add a unit test to check that either `LambdaruntimeError.connectionToControlPlaneLost`, a `ChannelError`, or an `IOError` is thrown when the server closes the connection

================================================
The below is the original PR description. It is kept here for history.

This PR implements a mechanism to propagate connection loss information from the Lambda runtime client to the runtime loop using `EventLoopFuture`, enabling termination without backtrace when the connection to the Lambda control plane (or a Mock Server) is lost.

This PR fixes https://github.com/swift-server/swift-aws-lambda-runtime/issues/465

**Changes Made**
Core Changes:

- `LambdaRuntimeClient`: Added `futureConnectionClosed` property to signal connection loss via `EventLoopFuture`
- `Lambda.swift`: Modified `runLoop` to monitor connection status and throw an error on connection loss
- `LambdaRuntime` and `ServiceLifecycle` Integration: Modified to properly handle connection loss (and other top-level) errors and avoid application to crash with a backtrace.

**Termination without a backtrace**

- When connection to Lambda control plane is lost, the runtime exits cleanly
- Prevents hanging or zombie processes in testing/development environments
- Maintains compatibility with AWS Lambda service behavior (new runtime environment creation)

**Use Cases**
- Performance Testing: Prevents hanging when testing against MockServer
- Development: Cleaner shutdown behavior during local testing
- Production: Aligns with AWS Lambda service expectations for connection loss scenarios

adam-fowler reviewed (commented):
I think you'd be better to use the state machine to work out state, instead of adding additional variables, outside of the state machine.

Why don't you just check for the disconnected state in `LambdaRuntimeClient.nextInvocation` and if it is set throw a `LambdaRuntimeError(code: .connectionToControlPlaneLost)`

sebsto commented:
> I think you'd be better to use the state machine to work out state, instead of adding additional variables, outside of the state machine.

That doesn't work in this context because I need to propagate the error from a callback inside `LambdaRuntimeClient` up to `Lambda`. The states are `private` to the `LambdaRuntimeClient` and the `LambdaChannelHandler` and can not be accessible from `Lambda` 

Throwing errors is not an option neither because the only place where I can trap the `connectionClosed` event while the client is blocked on `GET /next` is inside a callback `channel.closeFuture.whenComplete { result ... }` This block can not throw.

sebsto commented:
Because `connectionState` is private and making it `package` obliges to open a lot of internal types (such as `ConnectionState`, the `ChannelDelegate`,...). It also obliges to remove the `@UsableFromInline` on `Lambda.Run()` and dance with the actor isolation.  I'll send a modification to let see you the amount of changes that are needed (unless I miss something)

sebsto commented:
Here is a new version. 

- I introduced a new `connectionState` : `lostConnection` (because .disconnected also means "not yet connected") 
- I moved the check back to `Lambda` because if we do the check in `LambdaRuntimeClient.nextInvocation()` the method still hands on lost connections

sebsto commented:
Here is a new version.

 - I introduced a new `connectionState : lostConnection` (because .disconnected also means "not yet connected")
 - The check on connection lost is on `LambdaClientRuntime.nextInvocation()` 
 - When the connection is lost, the call back now correctly `resume(throwing:)` pending continuation, eliminating the hangs I was observing from time to time.
 - DRY : I moved the top-level error handling on `LambdaRuntime._run()` method instead of the two `run()` functions (one for service lifecycle and one regular)

sebsto commented:
I further simplified by removing a necessary extra state I introduced in the previous commit.  
I removed other unnecessary code to keep this PR as short as possible.  

Core changes are 
- LambdaRuntimeClient: lines 366 - 378
- LambdaChannelHandler: line 900 
- LambdaRuntime.swift: lines 109 - 119 

23 lines of code.
The rest are empty lines and `logger` statements.

adam-fowler commented on Sources/AWSLambdaRuntime/LambdaRuntimeClient.swift:
```diff
@@ -363,7 +363,19 @@ final actor LambdaRuntimeClient: LambdaRuntimeClientProtocol {
+
+                    // resume any pending continuation on the handler
+                    if case .connected(_, let handler) = runtimeClient.connectionState {
+                        if case .connected(_, let lambdaState) = handler.state {
```
This shouldn't be necessary, as `LambdaChannelHandler.channelInactive` will have resumed the continuation (with error `ioOnClosedChannel`). None of the tests trigger this code either.

sebsto replied:
`LambdaChannelHandlerHandler.channelInnactive` is indeed called when the server closes a connection but never in the state `waitingForNextInvocation`.

When the server closes the connection, the state in the `LambdaChannelHandler` is 
`connected` and `LambdaState.idle` 

To reproduce what I observe:

1. `while true; do MAX_INVOCATIONS=1 MODE=json PORT=7777 LOG_LEVEL=trace swift run MockServer; sleep 1; done`
2. from another terminal: `cd Examples/HelloJSON ; LAMBDA_USE_LOCAL_DEPS=../.. LOG_LEVEL=trace AWS_LAMBDA_RUNTIME_API=127.0.0.1:7777 swift run`

You'll that when the server closes the connection, `LambdaChannelHandler.channelInactive()` is called with `self state == .connected(_, .idle)`

sebsto replied:
Done, https://github.com/swift-server/swift-aws-lambda-runtime/pull/545/commits/b0234f5c3f576464c52caf9c1fe12702cc8e6a9b

sebsto replied:
As you noticed, the `connnectionToControlPlaneLost` error was not triggered by the previous commit.  This commit adds a test to make sure the two types of errors are triggered https://github.com/swift-server/swift-aws-lambda-runtime/pull/545/commits/156e6c7265346b5e1177293ffaf8b8130685a88d

sebsto replied:
Hello @adam-fowler can you check the updated test? It covers the two behaviours I’m observing (connection lost in `.idle` state and `.waitingForNextInvocation` state.

sebsto replied:
@adam-fowler Good catch.  I moved the promise fulfilment code to `channelInactive` using the code you suggested.  You're correct - this addresses the root cause

sebsto commented:
In French, we say "la nuit porte conseil" aka "good advices come during the nigh"
I refactored the test to use a timeout and a `CancellationError`
